pjlib/src/pjlib-test/ioq_udp.c

```diff
@@ -517,6 +510,46 @@ static int unregister_test(pj_bool_t allow_concur)
 	return -210;
     }
 
+    /* Now unregister again, immediately and after PJ_IOQUEUE_KEY_FREE_DELAY.
+     * It should return error, and most importantly, it must not crash.. */
+    for (i=0; i<2; ++i) {
+	status = pj_ioqueue_unregister(key);
+	/*
+	 * as it turns out, double unregistration returns PJ_SUCCESS
+	if (status == PJ_SUCCESS) {
+	    PJ_LOG(1, (THIS_FILE,
+		   "Expecting pj_ioqueue_unregister() error (i=%d)", i));
+	    return -220;
+	}
+	*/
+	pj_thread_sleep(PJ_IOQUEUE_KEY_FREE_DELAY + 100);
+    }
+
+    /*
+     * Second stage of the test. Register another socket. Then unregister using
+     * the previous key.
+     */
+    status = app_socket(pj_AF_INET(), pj_SOCK_DGRAM(), 0, RPORT, &rsock2);
+    if (status != PJ_SUCCESS) {
+	app_perror("Error initializing socket (2)", status);
+	return -330;
+    }
+
+    /* Register rsock to ioqueue. */
+    status = pj_ioqueue_register_sock(pool, ioqueue, rsock2, user_data2,
+				      &cb, &key2);
+    if (status != PJ_SUCCESS) {
+	app_perror("Error registering to ioqueue (2)", status);
+	return -340;
+    }
+
+
+    /* We shouldn't be able to unregister using the first key */
+    status = pj_ioqueue_unregister(key);
+    if (status == PJ_SUCCESS) {
+	PJ_LOG(2,(THIS_FILE, "....Warning: unregistering dead key was successful"));
```

Comment:
The test scenario is as follows:

1. `ioqueue_create(**max_handles=1** ) // <-- note: max_handles is just one` 
2. `key1 = ioqueue_register(sock1)`
3. `ioqueue_unregister(key1)`
4. `key2 = ioqueue_register(sock2)`
5. `status = ioqueue_unregister(key1)  // <-- note: unregister key1 again`

So in step 5, basically we're simulating buggy application that is unregistering an old key.

With current ioqueue implementation, it will return success because `key1` is the same as `key2` (due to SAFE_UNREGISTRATION in ioqueue). But what is the expected status in step 5 anyway? Should it return an error? Ideally, I think so. But since we can't do that, I'm putting this as a warning to remind us about this behavior. Or should we just ignore it (set verbosity to info/debug)?

Reply:
Just add some explanation in the comments should be fine, IMO, to avoid confusion in the future when we happen to revisit this again.

Reply:
done